Document certification issuing process

mbuechse opened the pull request:
Resolves #337 and #346.
Supersedes #239.

mbuechse commented:
See also the [rendered version](https://github.com/SovereignCloudStack/standards/blob/issue/337/Standards/scs-xxxx-v1-achieving-certification.md)

josephineSei commented on Standards/scs-xxxx-v1-achieving-certification.md:
```diff
@@ -0,0 +1,45 @@
+## Open Questions
+
+## Related Documents
+
```
Could you please link the other certificate scopes? Or a document describing them all?

mbuechse replied:
I added language to explain that this will be added once it becomes relevant. I think it's unnecessary to talk about these things because at this point in time, we (well, I) have only very vague ideas about them.

josephineSei commented on Standards/scs-xxxx-v1-achieving-certification.md:
```diff
@@ -0,0 +1,45 @@
+
+1. Each certificate issued pertains to a given cloud, a given scope, and a given version of that scope with a fixed expiry date. The certificate is only valid for that cloud and for the time frame that ends on that expiry date.
+
+2. The operator MUST include the official SCS compliance test suite (which does not require admin privileges) in their continuous test infrastructure (e.g., Zuul). The tests MUST be run at least nightly. For public clouds, it is recommended to offer the SCS project access to the infrastructure (ideally, including a Zuul nodepool) so the test suite runs can be triggered continuously by the SCS tea ...
```
Imho it should be discussed, whether tests should run on a daily basis. Some of them might have more impact to a deployment then other. (I am thinking about the domain manager role tests by @markus-hentsch ). For IaaS I see the necessity to test whenever an update is done in the deployment. For KaaS that might be different.

mbuechse replied:
Good point. We should partition the tests into quick ones (which should run nightly) and less quick ones.

However, I'm not sure if I understand what you mean by an update in the deployment. Someone can add a new flavor ad hoc without triggering any tests, right? And in the past, we had cases where suddenly a new flavor (that violated the specs) appeared in some cloud and then the nightly failed.

anjastrunk replied:
In SIG meeting on 19.10.23, we decided to group tests into different buckets (such as light, medium, heavy) This should be reflected here.

mbuechse replied:
I added the classification. I will have to update scs-0003 as well (mainly to document the additional field in the yaml, but without repeating the intervals).

josephineSei commented on Standards/scs-xxxx-v1-achieving-certification.md:
```diff
@@ -0,0 +1,45 @@
+
+   <!-- Initially this will probably be eMail -->
+
+3. If the desired certificate requires manual checks, then the operator MUST offer the SCS project suitable access. Manual checks SHOULD be repeated once every 3 months.
```
This is a huge difference between the daily (or nightly) automated tests and those only done every three months. I also think for IaaS at least such tests need to be done whenever a deployment is updated (e.g. config change, security fix, etc..).

mbuechse replied:
The constraint here is the capacity of the manual testers. Let's talk about this in a meeting, as you proposed, because it would be good for me to have a better understanding of the deployments you are referring to and how often this would happen. If it's 4 times a year, but not necessarily spaced out evenly, well, then it might work.

anjastrunk replied:
In SIG meeting on 19.10.23, we decided to run manual tests once per quarter.

mbuechse replied:
There is a small difference between "once every three months" (as suggested in the meeting of 2023-09-21) and "once per quarter" (as suggested in the meeting of 2023-10-20), because the former case implies the time distance between the tests is more or less always three months, whereas in the latter case the distance could also be almost zero (if we test near the end of one quarter and then near the beginning of the next quarter) or almost six months (if we test near the beginning of one quarter and near the end of the next). I would like to keep the former case here and let the community speak.

frosty-geek replied:
IMHO the "SHOULD" case is too weak and I would rather swap this for "Manual checks MUST be repeated at least every 6 month" give that the SCS Project has a 6 month release cycle.

mbuechse replied:
Added this question to the next meeting's agenda. https://input.scs.community/2023-scs-sig-standardization

mbuechse replied:
Changed the SHOULD into a MUST. Somehow the interval did not get changed, but I think having a "halftime inspection" is not too bad, actually.

josephineSei commented on Standards/scs-xxxx-v1-achieving-certification.md:
```diff
@@ -0,0 +1,45 @@
+
+7. If the certificate is to be revoked for any reason, it will be included in a publicly available Certificate Revokation List (CRL). This fact will also be reflected in the certificate status page.
+
+8. If any of the automated tests or manual checks fail after the certificate has been issued, the certificate is not immediately revoked. Rather, the automated tests MUST pass 99.x % of the runs, and the operator SHALL be notified at the second failed attempt in a row at the latest. In case a manual check fails, it has to be repeated at a date to be negotiated with SCS. It MAY NOT fail more than  ...
```
Good thing to have time to fix things :)

josephineSei reviewed (commented):
I only have one bigger point, that maybe should be discussed in a meeting. Its about the interval of the tests repeated.

neuroserve commented:
At point two I'd expect a link/reference to the (valid/versioned) "SCS compliance test suite". Same for the basis of the "classification" (e. g. what is meant by "light", "medium" and "heavy").

frosty-geek reviewed (approved):
added my 2¢ but overall LGTM either way

berendt commented on Standards/scs-xxxx-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+    - _medium_: at least weekly,
+    - _heavy_: at least monthly.
+
+   For public clouds, it is recommended to offer the SCS project access to the infrastructure (ideally, including a Zuul nodepool) so the test suite runs can be triggered continuously by the SCS team.
```
I think the setup of a dedicated Zuul nodepool is not required to run the test suite. It should be sufficient to provide a project that can be accessed from the Zuul system.

mbuechse replied:
Added this question to the next meeting's agenda. https://input.scs.community/2023-scs-sig-standardization

berendt commented on Standards/scs-xxxx-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+
+2. The operator MUST include the official SCS compliance test suite (which does not require admin privileges) in their continuous test infrastructure (e.g., Zuul). The tests MUST be run at given intervals, depending on their classification:
+
+    - _light_: at least nightly,
```
I would think of it the other way round? Heavy for Nightly and Light for Monthly?

mbuechse replied:
The classification is with regards to resource usage. We want the possibility to run resource-heavy checks less often. (If a check is critical, it might be put into the light bucket anyhow)

mbuechse replied:
Sorry, I meant to say that I wanted to add this. That got lost somehow while I edited my comment. So, yes, will do!

berendt commented on Standards/scs-xxxx-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+5. The SCS certification assessment body (or entities empowered to do so) WILL review the certification application and either grant the certification, reject it or ask for further measures or information.
+   <!-- body might as well just be a machine, at least on the scs compatible level -->
+
+6. Once the certificate is granted by the SCS certification assessment body, the operator MAY use the corresponding logo and publicly state the certified "SCS compatibility" on the respective layer for the time of the validity of the certification. The logo MUST be accompanied by a hyperlink (a QR code for printed assets) to the respective certificate status page.
```
Wouldn't it make more sense to demand visibility from the CSP if we put in the effort to do the tests etc?

mbuechse replied:
Added this question to the next meeting's agenda. https://input.scs.community/2023-scs-sig-standardization

anjastrunk commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+
+## Regulations
+
+1. Each certificate issued pertains to a given cloud, a given scope, and a given version of that scope with a fixed expiry date. The certificate is only valid for that cloud and for the time frame that ends on that expiry date.
```
As each certificate version has an expiry date and each certificate has an expiry date, would it makes sense, to add a sentence how these dates interrelate? Do we support certificate expiry dates that are later in time than the expiry date of the corresponding certificate version? Or are certificates maximum valid until the expiry date of certificates' version?

mbuechse replied:
Good point. I always assumed it was self-evident that the certificate expires at the same date as the version of the scope; however, it could also expire earlier. So it does make sense to make this explicit! I think we agreed that the certificate should not expire earlier, right?

mbuechse replied:
On second thought, the sentence already states precisely that:
> The certificate is valid ... for the time frame that ends on that expiry date.

mbuechse replied:
Marking this as resolved because @fdobrovolny raised a similar question. Let's address it there.

anjastrunk commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+
+3. If the desired certificate requires manual checks, then the operator MUST offer the SCS project suitable access. Manual checks MUST be repeated once every 3 months.
+
+4. Details on the standards achieved, as well as the current state and the history of all test and check results will be displayed on a public webpage (henceforth, _certificate status page_) owned by SCS.
```
As we publish results of SCS conformance tests, we need the permission of CSP to do so in order to prevent any legally conficrts. I do not know, if this should be part of "Regulations for achieving SCS-compatible certification". However, we should keep this in mind and add an appropriate statement to SCS certificate's terms and conditions.

mbuechse replied:
My understanding is that any CSP objecting to this kind of publication should NOT apply for certification. Nobody is being forced.

garloff replied:
I guess we need to write a little "SCS certification terms & conditions document" that states obligations and conditions. I don't to sign individual contracts, so it's a T&C thing. (In German: Allgemeine Geschäftsbedingungen.)

anjastrunk commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+
+4. Details on the standards achieved, as well as the current state and the history of all test and check results will be displayed on a public webpage (henceforth, _certificate status page_) owned by SCS.
+
+5. The SCS certification assessment body (or entities empowered to do so) WILL review the certification application and either grant the certification, reject it or ask for further measures or information.
```
In favor of transparency, we should make "SCS certificate assessment body" explicitly. I.e. who will issue which type of certificate.

garloff replied:
Currently, the SCS project (with the legal entity OSBA behind it) is the only "SCS certification assessment body", but we may indeed empower others in the future to do this on our behalf.

anjastrunk commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+5. The SCS certification assessment body (or entities empowered to do so) WILL review the certification application and either grant the certification, reject it or ask for further measures or information.
+   <!-- body might as well just be a machine, at least on the scs compatible level -->
+
+6. Once the certificate is granted by the SCS certification assessment body, the operator SHOULD use the corresponding logo and publicly state the certified "SCS compatibility" on the respective layer for the time of the validity of the certification. In case of a public cloud, this public display is even REQUIRED. In any case, the logo MUST be accompanied by a hyperlink (a QR code for printed as ...
```
Where do CSP download certificate's logo? Is the logot part of of the certificate? 

Follow up issue: We should define the technical implementation of temper-evident SCS certificates, especially, if they are digital, which I assume.

mbuechse replied:
@maxwolfs Do you know what file we can link to for a logo?

Regarding the tamper-proof badges: so far, people are advised to check the SCS-owned status page to check whether the certificate is still valid. After all, it can be revoked, so looking at the certificate itself (even if cryptographically signed) is not enough. As far as I recall, Kurt said the whole security thing is of lower priority for the time being.

anjastrunk replied:
You are right. I remember. We can go with the web site for now.

anjastrunk reviewed (changes requested):
See comments inline.

garloff commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+
+4. Details on the standards achieved, as well as the current state and the history of all test and check results will be displayed on a public webpage (henceforth, _certificate status page_) owned by SCS.
+
+5. The SCS certification assessment body (or entities empowered to do so) WILL review the certification application and either grant the certification, reject it or ask for further measures or information.
```
```suggestion
5. The SCS certification assessment body (initially the SCS project in the OSB Allliance e.V., possibly further entities empowered to do so by the SCS trademark owner, currently the OSB Alliance e.V.) WILL review the certification application and either grant the certification, reject it or ask for further measures or information.
```

garloff commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,55 @@
+
+   <!-- Initially this will probably be eMail -->
+
+3. If the desired certificate requires manual checks, then the operator MUST offer the SCS project suitable access. Manual checks MUST be repeated once every 3 months.
```
```suggestion
3. If the desired certificate requires manual checks, then the operator MUST offer the SCS project suitable access. Manual checks MUST be repeated once every quarter.
```

garloff replied:
This was discussed before. Once per quarter gives us a bit more flexibility to schedule the manual checks for which we may have good reasons, such as pushing a manual check behind a change, e.g. a scheduled upgrade or just scheduling tester resources.

garloff reviewed (approved):
Overall looks good to me.
I have a slight preference for "once per quarter" over "once every 3 months" because of the stated reasons. Feel free to apply this proposed change.
Next steps then would be to merge this and ask for opinions before we turn the draft into an approved document.

fdobrovolny reviewed (approved):
I now saw that I did a review a week ago but did not send it.

I added my two cents other than that LGTM.

mbuechse commented:
> I now saw that I did a review a week ago but did not send it.
> 
> I added my two cents other than that LGTM.

Thanks @fdobrovolny -- very good points! I added them to today's agenda.

mbuechse commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,56 @@
+
+2. The operator MUST include the official [SCS compliance test suite](https://github.com/SovereignCloudStack/standards/tree/main/Tests) (which does not require admin privileges) in their continuous test infrastructure (e.g., Zuul). The tests MUST be run at given intervals, depending on their resource-usage classification:
+
+    - _light_: at least nightly,
```
It's been dealt with. The yaml file allows for the classification to be specified. So far, all scripts use the default value (light).

fkr commented:
I just talked to @mbuechse since I was missing the part of "things to check for compliance that requires system level access" - which will (of course) be done by the operator with a results file submitted to SCS.
@mbuechse and I agreed that we will open an issue for that and revisit that ONCE we have the first compliance check that requires this (since so far, we have none that require that).

fkr reviewed (approved):
Thanks @mbuechse for your efforts on this!

josephineSei commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,56 @@
+
+   <!-- Initially this will probably be eMail -->
+
+3. If the desired certificate requires manual checks, then the operator MUST offer the SCS project suitable access. Manual checks MUST be repeated once every quarter.
```
In an airgapped scenario this would mean a real person has to go their and do the checks. And even that is not always possible, as the physical entrance to data centers may be restricted to certain persons. (I have already seen this in public cloud offerings).

josephineSei commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,56 @@
+
+   For public clouds, it is recommended to offer the SCS project access to the infrastructure so the test suite runs can be triggered continuously by the SCS team.
+
+   Alternatively, and for non-public clouds, the results (log files) MUST be submitted to SCS by a mechanism of SCS' choice and need to be reproduced again on request by SCS.
```
We should have a set of options for providing the results ready. On some airgapped infrastructures getting information like that out of it is nearly impossible as every information is classified. That may even lead to a letter with printed logs, that need to be destroyed in certain ways after reading.

josephineSei commented on Standards/scs-0004-v1-achieving-certification.md:
```diff
@@ -0,0 +1,56 @@
+
+3. If the desired certificate requires manual checks, then the operator MUST offer the SCS project suitable access. Manual checks MUST be repeated once every quarter.
+
+4. Details on the standards achieved, as well as the current state and the history of all test and check results of the past 18 months will be displayed on a public webpage (henceforth, _certificate status page_) owned by SCS.
```
As for some infrastructures we will talk about classified information, it would be very interesting to define exactly, what information will be displayed. (No need to do this here in this standard, but it will be good to have such information.)

josephineSei reviewed (commented):
All my comments are of informational nature.

I put in a few implications this will have on airgapped and high-security infrastructures. 

To summarize: this might be enough for operators of such infrastructures to not go for scs-compatability. I have no strong opinion on whether this is okay or not.
